src/server.rs

```diff
@@ -271,6 +310,63 @@ impl Server {
         }
     }
 
+    // Server browser related query packet processing for shared socket mode.  These are used
+    // when you pass STEAMGAMESERVER_QUERY_PORT_SHARED as the query port to SteamGameServer_Init.
+    // IP address and port are in host order, i.e 127.0.0.1 == 0x7f000001
+
+    // These are used when you've elected to multiplex the game server's UDP socket
+    // rather than having the master server updater use its own sockets.
+    //
+    // Source games use this to simplify the job of the server admins, so they
+    // don't have to open up more ports on their firewalls.
+
+    // Call this when a packet that starts with 0xFFFFFFFF comes in. That means
+    // it's for us.
+    pub fn handle_incoming_packet(
```

Comment:
Here, data can be an immutable reference and size is `data.len()`. Instead of raw IP as u32
 and port, use `SocketAddrV4` from std.

```rs
pub fn handle_incoming_packet(data: &[u8], addr: SocketAddrV4)
```

Reply:
I'm not super familiar with rust, but as far as I can tell there's no way (in std) to provide a smaller view on an array without allocating for a slice/new array. That makes me think that there is some utility in having size be a separate value, to reuse the same data array across multiple calls to `handle_incoming_packet` in order to reduce allocations.

Reply:
You do it by "slicing" the existing slice, array, Vec etc. The syntax is indexing with a Range type, `big_array[2..15]`. In this case you'd write `&my_buf[..actual_len]` when passing in the data